Pull request re: Issue #7

dcunning11235 opened the pull request:
Addresses "Use full file if already downloaded and lite file is requested" https://github.com/dkirkby/bossdata/issues/7.

Implements:

| Lite DB | Lite File | Full DB | Full File | ACTION |
| --- | --- | --- | --- | --- |
| 1 | * | * | * | Use Lite DB |
| 0 | * | 1 | * | Use Full DB |
| 0 | 1 | 0 | * | Use Lite File/Build Lite DB |
| 0 | 0 | 0 | 1 | Use Full File/Build Full DB |
| 0 | 0 | 0 | 0 | Download Lite File/Build Lite DB |

dcunning11235 committed:
Synching changes so far; not a buildable version!

dcunning11235 committed:
Fix _db_path_helper signature; made static

dcunning11235 committed:
Completed changes to use Full DB or file is Lite DB or file is
requested, is not present, but the Full version is present.

dcunning11235 committed:
flake8 cleanup

dkirkby commented on bossdata/remote.py:
```diff
@@ -202,9 +202,18 @@ def get(self, remote_path, progress_min_size=10, auto_download=True):
+
+        # Don't test if its a string; maybe later it won't be.  Test if it's a list
+        remote_paths = []
+        if not isinstance(remote_path, list):
```
A tuple should also work here.  More generally, any iterable container is fine.  I believe you can identify any iterable using `hasattr(remote_path,__iter__)`.

dkirkby commented on bossdata/remote.py:
```diff
@@ -186,9 +186,9 @@ def get(self, remote_path, progress_min_size=10, auto_download=True):
+            remote_path(str,list): The full path to the remote file relative to the remote
                 server root, which should normally be obtained using :class:`bossdata.path`
-                methods.
+                methods; OR a list of such paths.
```
Need to explain how a list (or iterable) of paths will be used.

dkirkby commented on bossdata/remote.py:
```diff
@@ -212,6 +221,9 @@ def get(self, remote_path, progress_min_size=10, auto_download=True):
         # If we get here, the file is not available locally so try to download it now.
         # Create local directories as needed.
+
+        # We get the file for first path is there are more than one
```
typo: is -> if

dcunning11235 committed:
Added in 'full if no lite' chnages for bossplot

dcunning11235 committed:
Removed confusing comment

dcunning11235 commented:
Added in changes per comments.

Included bossplot changes re: using full spec file over lite spec file if full version is locally available but lite requires download.

dkirkby commented:
I am testing the meta part of this now with:

```
bin/bossquery --where 'PLATE=6641' --max-rows 10 --print --verbose
```

under the following conditions:

```
0. All four meta files are available.
1. mv spAll-v5_7_0-lite.db spAll-v5_7_0-lite.db-hidden
2. mv spAll-v5_7_0.db spAll-v5_7_0.db-hidden
3. mv spAll-v5_7_0.dat.gz spAll-v5_7_0.dat.gz-hidden
   rm spAll-v5_7_0-lite.db # since it was recreated by 2.
4. mv spAll-v5_7_0.fits spAll-v5_7_0.fits-hidden
    rm spAll-v5_7_0.db # since it was recreated by 3.
```

Here are the results:
- Tests 0 and 1 both use the available lite db, as expected.
- Test 2 builds the lite db from the available .dat.gz file, as expected.
- Test 3 builds the full db from the available .fits file, as expected.
- Test 4 downloads the .dat.gz file, as expected.
  This all looks good!  One thing I noticed (since I interrupted `create_meta_lite` with ^C) is that we have no protection against a partial build of a sqlite db.  I just created #45 to deal with this.

dkirkby commented:
Last items before we close this out:
- Mention when `bossquery` or `bossplot` is doing a substitution if `--verbose` is specified (assuming that this is relatively easy to do).
- Update the `bin/scripts.rst` docs to mention the new functionality.

dcunning11235 committed:
Merge pull request #2 from dkirkby/master

Pull changes from main to my fork

dcunning11235 committed:
Merge branch 'master' down into remote__get

In addtion, new changes:
bossplot:  added messaging (when --verbose) to when using full files because lite not available.
bossquery:  added messageing (when --verbose) when using full DB because lite not available.
meta.py:  added 'lite' attribute to Database, indicates which type is being accessed
remote.py:  altered get to return (str, int) rather than str; the int indicates the index of the
	remote path passed as argument that was used or 0 if only a string was passed.

Conflicts:
	bin/bossplot

dkirkby commented on bossdata/remote.py:
```diff
@@ -197,14 +200,30 @@ def get(self, remote_path, progress_min_size=10, auto_download=True):
 
         Returns:
-            str: Absolute local path of the local file that mirrors the remote file.
+            str, int: Absolute local path of the local file that mirrors the remote file, and
```
Adding the index to the return value forces all API users to handle it, even though its usually an implementation detail they don't care about.  It also breaks all code that uses `get` that has not yet been updated (e.g., `bossfetch`).  Instead, we can let the caller that actually cares about this use something like:

```
local_path = mirror.get(remote_paths)
if local_path != remote_paths[0]:
    print('{} was substituted for {}.'.format(local_path, remote_paths[0]))
```

This pattern should be documented in the docstring for `get`.

dkirkby replied:
I just realized why my suggested pattern for detecting substitutions by get() doesn't work (because remote_paths[0] isn't a local path!) but I just update the get() logic to fix this in cf610b7de487f77834.

dkirkby replied:
I actually made things worse but now they should finally be better in f8a0d8a.

dkirkby commented on bin/bossplot:
```diff
@@ -121,13 +121,20 @@ def main():
             if args.verbose:
                 print('Exposure summary:')
                 print(specfile.exposure_table)
+                if lite and ind > 0:
```
`if local_path != remote_paths[0]:`

dkirkby commented on bin/bossplot:
```diff
@@ -121,13 +121,20 @@ def main():
+                remote_paths.append(finder.get_spec_path(plate=args.plate, mjd=args.mjd,
+                                fiber=args.fiber, lite=False))
+
+            local_path, ind = mirror.get(remote_paths)
```
`local_path = mirror.get(remote_paths)`

dkirkby commented on bin/bossquery:
```diff
@@ -44,6 +44,8 @@ def main():
 
     if args.verbose:
+        if not args.full and not meta_db.lite:
+            print("Using --full databse as lite files were not available.")
```
typo

dkirkby commented:
@dcunning11235 For your next issue, you might find it easier to create a branch directly in this repo rather than working on a clone in your own repo, now that you are a registered collaborator.

dcunning11235 committed:
Removed tuple return from remote.Manager.get, restoring it to previous signature/return val
Added (longish) doc-string describing how to determine which remote path was returned for
Cleanup for in bossplot for this change/reversion, addition of code from docstring

dcunning11235 commented:
@dkirkby Updated per your comments

dkirkby commented:
I just ran the following tests:
- `make html` in `docs/`
- remove lite file and check that full file is used instead
- remove lite and full files and check that lite file is downloaded

Looks good!
